worker/provisioner/provisioner_task.go

```diff
@@ -689,106 +714,357 @@ func (task *provisionerTask) maintainMachines(machines []*apiprovisioner.Machine
+			continue
 		}
+		instanceMachines[instId] = machine.Id()
+	}
 
-		instanceCfg, err := task.constructInstanceConfig(m, task.auth, pInfo)
-		if err != nil {
-			return task.setErrorStatus("creating instance config for machine %q: %v", m, err)
+	// convert instances IDs to machines IDs to aid distributing
+	// not yet created instances across availability zones.
+	task.availabilityZoneMachines = make([]*availabilityZoneMachine, len(availabilityZoneInstances))
+	for i, instances := range availabilityZoneInstances {
+		machineIds := set.NewStrings()
+		for _, instanceId := range instances.Instances {
+			if id, ok := instanceMachines[instanceId]; ok {
+				machineIds.Add(id)
+			}
 		}
+		task.availabilityZoneMachines[i] = &availabilityZoneMachine{
+			zoneName:         instances.ZoneName,
+			machineIds:       machineIds,
+			failedMachineIds: set.NewStrings(),
+		}
+	}
+	return nil
+}
 
-		assocProvInfoAndMachCfg(pInfo, instanceCfg)
+// populateDistributionGroupZoneMap returns a zone mapping which only includes
+// machines in the same distribution group.  This is used to determine where new
+// machines in that distribution group should be placed.
+func (task *provisionerTask) populateDistributionGroupZoneMap(machineIds []string) []*availabilityZoneMachine {
+	var dgAvailabilityZoneMachines []*availabilityZoneMachine
+	dgSet := set.NewStrings(machineIds...)
+	for _, azm := range task.availabilityZoneMachines {
+		dgAvailabilityZoneMachines = append(dgAvailabilityZoneMachines, &availabilityZoneMachine{
+			azm.zoneName,
+			azm.machineIds.Intersection(dgSet),
+			azm.failedMachineIds,
+		})
+	}
+	return dgAvailabilityZoneMachines
+}
 
-		var arch string
-		if pInfo.Constraints.Arch != nil {
-			arch = *pInfo.Constraints.Arch
-		}
+// machineAvailabilityZoneDistribution returns a map of the specified machines
+// to a suggested availability zone to start in.  Machines are spread across
+// availability zones based on lowest population of the "available" zones.
+// Machines in the same DistributionGroup are placed in different zones, spread
+// across availability zones based on lowest population of machines in that
+// DistributionGroup.
+func (task *provisionerTask) machineAvailabilityZoneDistribution(machines []*apiprovisioner.Machine) (map[*apiprovisioner.Machine]string, error) {
+	machineZones := make(map[*apiprovisioner.Machine]string, len(machines))
+
+	if len(machines) == 0 {
+		return machineZones, nil
+	}
 
-		v, err := m.ModelAgentVersion()
-		if err != nil {
-			return errors.Trace(err)
+	// Get existing machine distributions.
+	if err := task.populateAvailabilityZoneMachines(); err != nil {
+		return machineZones, err
+	}
+	// Retrieve any DistributionGroups for these machines
+	dgMachines := make([]names.MachineTag, len(machines))
+	for i, machine := range machines {
+		dgMachines[i] = machine.MachineTag()
+	}
+	distributionGroups, err := task.distributionGroupFinder.DistributionGroupByMachineId(dgMachines...)
```

Comment:
We're calling machineAvailabilityZoneDistribution for each successively attempted availability zone, but the distribution group shouldn't have changed in any relevant way. Machines might have been added, but they won't be added to the availability zone machine map until they're presented to startMachines. So it would be great if we could avoid making repeat calls to this.

Maybe have startMachines call DistributionGroupByMachineId, then pass the result into machineAvailabilityZoneDistribution and startMachine (so the latter can pass it into the former).

Reply:
done.